src/compiler/checker.ts

```diff
@@ -4722,6 +4846,1104 @@ namespace ts {
+                            statements = filter(statements, s => s !== exportDecl);
+                        }
+                        else {
+                            // some items filtered, others not - update the export declaration
+                            // (mutating because why not, we're building a whole new tree here anyway)
+                            exportDecl.exportClause!.elements = createNodeArray(replacements);
+                        }
+                    }
+
+                    // Not a cleanup, but as a final step: If there is a mix of `export` and non-`export` declarations, but no `export =` or `export {}` add a `export {};` so
+                    // declaration privacy is respected.
+                    if (enclosingDeclaration &&
+                        ((isSourceFile(enclosingDeclaration) && isExternalOrCommonJsModule(enclosingDeclaration)) || isModuleDeclaration(enclosingDeclaration)) &&
+                        (!some(statements, isExternalModuleIndicator) || (!hasScopeMarker(statements) && some(statements, needsScopeMarker)))) {
+                        statements.push(createEmptyExports());
+                    }
+                    return statements;
+                }
+
+                function canHaveExportModifier(node: Statement) {
+                    return isEnumDeclaration(node) ||
+                            isVariableStatement(node) ||
+                            isFunctionDeclaration(node) ||
+                            isClassDeclaration(node) ||
+                            (isModuleDeclaration(node) && !isExternalModuleAugmentation(node) && !isGlobalScopeAugmentation(node)) ||
+                            isInterfaceDeclaration(node) ||
+                            isTypeDeclaration(node);
+                }
+
+                function addExportModifier(statement: Statement) {
+                    const flags = (getModifierFlags(statement) | ModifierFlags.Export) & ~ModifierFlags.Ambient;
+                    statement.modifiers = createNodeArray(createModifiersFromModifierFlags(flags));
+                    statement.modifierFlagsCache = 0;
+                }
+
+                function visitSymbolTable(symbolTable: SymbolTable, suppressNewPrivateContext?: boolean, propertyAsAlias?: boolean) {
+                    const oldDeferredList = deferredPrivates;
+                    if (!suppressNewPrivateContext) {
+                        deferredPrivates = createMap();
+                    }
+                    symbolTable.forEach((symbol: Symbol) => {
+                        serializeSymbol(symbol, /*isPrivate*/ false, !!propertyAsAlias);
+                    });
+                    if (!suppressNewPrivateContext) {
+                        // deferredPrivates will be filled up by visiting the symbol table
+                        // And will continue to iterate as elements are added while visited `deferredPrivates`

```

Comment:
huzzah!

...is this true even on ES5 using our polyfill?

Reply:
Yep - we had a community member fix that a long time ago, since it not doing that (like two years ago) also broke the playground in ie.